### PR TITLE
Storage: Fixed typo in lvm.thinpool setting, adds patch to convert to lvm.thinpool_name as node specific key

### PR DESCRIPTION
**Pull request by AndrewElvisDeng**
**lxd/db/storage_pools.go:**
renamed 'lvm.thinpool' to 'lvm.thinpool_name'

**lxd/patches.go**
created a patch to rename existing, incorrectly set configs from the typo

Signed-off-by: Andrew Deng <adeng1433@gmail.com>

### CODE REVIEW

File: lxd/patches.go [hunk 1]
```diff
@@ -3714,3 +3713,65 @@ func patchUpdateFromV30(_ *sql.Tx) error {
...
+			_, err := tx.Exec(`
+			INSERT INTO storage_pools_config(storage_pool_id, node_id, key, value)
+		  	VALUES(?, ?, 'lvm.thinpool_name', ?)
+			`, poolID, curNodeID, value)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create %s node config", key)
+			}
+		}
+	}
+	}
```
Comment (stgraber) [inline lxd/patches.go:3768]: That looks suspicious

File: lxd/patches.go [hunk 2]
```diff
@@ -3714,3 +3713,65 @@ func patchUpdateFromV30(_ *sql.Tx) error {
...
+		
+	err = tx.Commit()
+	if err != nil {
+		return errors.Wrap(err, "failed to commit transaction")
+	}
+		
+	return err
+
```
Comment (stgraber) [inline lxd/patches.go:3776]: Can drop the extra line break

File: lxd/patches.go [hunk 3]
```diff
@@ -3714,3 +3713,65 @@ func patchUpdateFromV30(_ *sql.Tx) error {
 
 	return nil
 }
+
+//  renames any config incorrectly set config file entries due to the lvm.thinpool_name typo
+func patchThinpoolTypoFix(name string, d *Daemon) error {
+	tx, err := d.cluster.Begin()
```
Comment (stgraber) [inline lxd/patches.go:3719]: Instead of manually creating a transaction that way, the entire part of the function interacting with the database should be in a `err = d.cluster.Transaction(func(tx *db.ClusterTx) error {` type block.

That way things will properly revert on failure rather than potentially keeping the database in a locked state.

**Comment by stgraber** [comment]
Left a few initial comments. I'm off tomorrow but @tomponline may be able to help with a more complete review.

File: lxd/patches.go [hunk 4]
```diff
@@ -163,8 +164,6 @@ func patchesApply(d *Daemon, stage patchStage) error {
 	return nil
 }
 
-// Patches begin here
```
Comment (tomponline) [inline lxd/patches.go:166]: @AndrewElvisDeng why did you remove this?

File: lxd/patches.go [hunk 5]
```diff
@@ -3714,3 +3713,65 @@ func patchUpdateFromV30(_ *sql.Tx) error {
 
 	return nil
 }
+
+//  renames any config incorrectly set config file entries due to the lvm.thinpool_name typo
+func patchThinpoolTypoFix(name string, d *Daemon) error {
```
Comment (tomponline) [inline lxd/patches.go:3718]: @AndrewElvisDeng please can you move this function above `patchMoveBackupsInstances` so that it is the most recently added patch (they are added to the file in reverse order). Thanks

File: lxd/patches.go (hunk 1 above)
Comment (tomponline) [reply lxd/patches.go:3768]: @AndrewElvisDeng please ensure that the code has been run `go fmt` before committing it. Thanks

File: lxd/patches.go [hunk 6]
```diff
@@ -3714,3 +3713,65 @@ func patchUpdateFromV30(_ *sql.Tx) error {
 
 	return nil
 }
+
+//  renames any config incorrectly set config file entries due to the lvm.thinpool_name typo
```
Comment (tomponline) [inline lxd/patches.go:3717]: @AndrewElvisDeng just a nit, but its not really a `config file` as such :)

**Review by tomponline — APPROVED** [review]
Apart from some minor issues with code ordering and formatting, this looks great. I can't see anything that looks to be a problem. Can you confirm that you've tested this patch on a cluster with that old config key set and you've seen it successfully replace it with a new config key for each node? Thanks

**Review by tomponline — CHANGES_REQUESTED** [review]
See discussion above.

**Comment by tomponline** [comment]
@AndrewElvisDeng I've tweaked the PR title to be a bit more descriptive also.

**Comment by AndrewElvisDeng** [comment]
@tomponline We've made the fixes you've requested, but we're struggling to figure out how to change the config file values specifically to recreate the original bug. Can you point us in the right direction?

File: lxd/patches.go [hunk 7]
```diff
@@ -164,6 +165,47 @@ func patchesApply(d *Daemon, stage patchStage) error {
 }
 
 // Patches begin here
+// renames any config incorrectly set entries due to the lvm.thinpool_name typo
+func patchThinpoolTypoFix(name string, d *Daemon) error {
+
```
Comment (stgraber) [inline lxd/patches.go:170]: Needles blank line (would cause a static analysis failure)

File: lxd/patches.go [hunk 8]
```diff
@@ -164,6 +165,47 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+	tx, err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+
```
Comment (stgraber) [inline lxd/patches.go:172]: Needles blank line (would cause a static analysis failure)

File: lxd/patches.go [hunk 9]
```diff
@@ -164,6 +165,47 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+		// Fetch the IDs of all existing nodes.
+		nodeIDs, err := query.SelectIntegers(tx, "SELECT id FROM nodes")
```
Comment (stgraber) [inline lxd/patches.go:174]: `err` left unchecked

File: lxd/patches.go [hunk 10]
```diff
@@ -164,6 +165,47 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+
+		// Fetch the IDs of all existing lvm pools.
+		poolIDs, err := query.SelectIntegers(tx, "SELECT id FROM storage_pools WHERE driver='lvm'")
```
Comment (stgraber) [inline lxd/patches.go:177]: `err` left unchecked

File: lxd/patches.go [hunk 11]
```diff
@@ -164,6 +165,47 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+
+		for _, poolID := range poolIDs {
+			// Fetch the config for this lvm pool and check if it has the
+			// lvn.thinpool_name key.
+			config, err := query.SelectConfig(
```
Comment (stgraber) [inline lxd/patches.go:182]: `err` left unchecked

File: lxd/patches.go [hunk 12]
```diff
@@ -164,6 +165,47 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+				tx, "storage_pools_config", "storage_pool_id=?", poolID)
+
+			value, ok := "lvm.thinpool"
```
Comment (stgraber) [inline lxd/patches.go:185]: What's this supposed to do?

Did you mean it to be `value, ok := config["lvm.thinpool"]` perhaps?

File: lxd/patches.go [hunk 13]
```diff
@@ -164,6 +165,47 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+			if !ok {
+				continue
+			}
+
+			// Delete the current key
+			_, err = tx.Exec(`
+				DELETE FROM storage_pools_config WHERE key='lvm.thinpool' AND storage_pool_id=?`, poolID)
```
Comment (stgraber) [inline lxd/patches.go:192]: `err` left unchecked

File: lxd/patches.go [hunk 14]
```diff
@@ -164,6 +165,47 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+
+			// Add the config entry for each node
+			for _, nodeID := range nodeIDs {
+				_, err := tx.Exec(`
```
Comment (stgraber) [inline lxd/patches.go:196]: `err` left unchecked

File: lxd/patches.go [hunk 15]
```diff
@@ -166,6 +167,58 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
 
+// renames any config incorrectly set entries due to the lvm.thinpool_name typo
+func patchThinpoolTypoFix(name string, d *Daemon) error {
+	tx, err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+
+		// Fetch the IDs of all existing nodes.
+		nodeIDs, err := query.SelectIntegers(tx, "SELECT id FROM nodes")
+		if err != nil {
+			return errors.Wrap(err, "failed to get IDs of current nodes")
+		}
```
Comment (stgraber) [inline lxd/patches.go:178]: Should put a line break after this one.

File: lxd/patches.go [hunk 16]
```diff
@@ -166,6 +167,58 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+		// Fetch the IDs of all existing lvm pools.
+		poolIDs, err := query.SelectIntegers(tx, "SELECT id FROM storage_pools WHERE driver='lvm'")
+		if err != nil {
+			return errors.Wrap(err, "failed to get IDs of current lvm pools")
+		}
```
Comment (stgraber) [inline lxd/patches.go:183]: Should put a line break after this one.

File: lxd/patches.go [hunk 17]
```diff
@@ -166,6 +167,58 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+		for _, poolID := range poolIDs {
+			// Fetch the config for this lvm pool and check if it has the
+			// lvn.thinpool_name key.
+			config, err := query.SelectConfig(
+				tx, "storage_pools_config", "storage_pool_id=?", poolID)
+			if err != nil {
+				return errors.Wrap(err, "failed a fetch of lvm pool config")
+			}
```
Comment (stgraber) [inline lxd/patches.go:191]: Should put a line break after this one.

File: lxd/patches.go [hunk 18]
```diff
@@ -166,6 +167,58 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+			if !ok {
+				continue
+			}
+
+			// Delete the current key
+			_, err = tx.Exec(`
+				DELETE FROM storage_pools_config WHERE key='lvm.thinpool' AND storage_pool_id=?`, poolID)
+			if err != nil {
+				return errors.Wrapf(err, "failed to delete %s config", key)
+			}
```
Comment (stgraber) [inline lxd/patches.go:202]: Should put a line break after this one.

File: lxd/patches.go [hunk 19]
```diff
@@ -166,6 +167,58 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
 // Patches begin here
 
+// renames any config incorrectly set entries due to the lvm.thinpool_name typo
+func patchThinpoolTypoFix(name string, d *Daemon) error {
+	tx, err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+
+		// Fetch the IDs of all existing nodes.
+		nodeIDs, err := query.SelectIntegers(tx, "SELECT id FROM nodes")
+		if err != nil {
+			return errors.Wrap(err, "failed to get IDs of current nodes")
```
Comment (stgraber) [inline lxd/patches.go:177]: LXD errors should start with a capital letter

File: lxd/patches.go [hunk 20]
```diff
@@ -166,6 +167,58 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+		}
+		// Fetch the IDs of all existing lvm pools.
+		poolIDs, err := query.SelectIntegers(tx, "SELECT id FROM storage_pools WHERE driver='lvm'")
+		if err != nil {
+			return errors.Wrap(err, "failed to get IDs of current lvm pools")
```
Comment (stgraber) [inline lxd/patches.go:182]: LXD errors should start with a capital letter

File: lxd/patches.go [hunk 21]
```diff
@@ -166,6 +167,58 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+		}
+		for _, poolID := range poolIDs {
+			// Fetch the config for this lvm pool and check if it has the
+			// lvn.thinpool_name key.
+			config, err := query.SelectConfig(
+				tx, "storage_pools_config", "storage_pool_id=?", poolID)
+			if err != nil {
+				return errors.Wrap(err, "failed a fetch of lvm pool config")
```
Comment (stgraber) [inline lxd/patches.go:190]: LXD errors should start with a capital letter

File: lxd/patches.go [hunk 22]
```diff
@@ -166,6 +167,58 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+			value, ok := config["lvm.thinpool"]
+			if !ok {
+				continue
+			}
+
+			// Delete the current key
+			_, err = tx.Exec(`
+				DELETE FROM storage_pools_config WHERE key='lvm.thinpool' AND storage_pool_id=?`, poolID)
+			if err != nil {
+				return errors.Wrapf(err, "failed to delete %s config", key)
```
Comment (stgraber) [inline lxd/patches.go:201]: LXD errors should start with a capital letter

File: lxd/patches.go [hunk 23]
```diff
@@ -166,6 +167,58 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+			}
+			// Add the config entry for each node
+			for _, nodeID := range nodeIDs {
+				_, err := tx.Exec(`
+				INSERT INTO storage_pools_config(storage_pool_id, node_id, key, value)
+				VALUES(?, ?, 'lvm.thinpool_name', ?)
+				`, poolID, curNodeID, value)
+				if err != nil {
+					return errors.Wrapf(err, "failed to create %s node config", key)
```
Comment (stgraber) [inline lxd/patches.go:210]: LXD errors should start with a capital letter

File: lxd/patches.go [hunk 24]
```diff
@@ -166,6 +167,58 @@ func patchesApply(d *Daemon, stage patchStage) error {
...
+				}
+			}
+		}
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to commit transaction")
```
Comment (stgraber) [inline lxd/patches.go:216]: LXD errors should start with a capital letter

File: lxd/patches.go (hunk 8 above)
Comment (stgraber) [reply lxd/patches.go:173]: still need to drop this blank line here

**Comment by comannnnndooooo** [comment]
Thanks for the prompt responses! We're fixing everything as you put it up!

**Comment by stgraber** [comment]
Thanks!

I just happened to be around. It's just our usual list of nitpicks and code styling really :)

**Comment by AndrewElvisDeng** [comment]
Ran into too many git issues: re-forked and created new pull request: #8039 . Closing this PR
